packages/cpp/src/common/cpp-preferences.ts

```diff
@@ -14,32 +14,38 @@ import {
     PreferenceSchema
 } from '@theia/preferences/lib/common';
 
+export const CLANGD_COMMAND_DEFAULT = 'clangd';
+
 export const CppConfigSchema: PreferenceSchema = {
     "type": "object",
     "properties": {
-        "cpp.clangdCompileCommandsPath": {
+        "cpp.clangdCompilationDatabaseDirectory": {
             "type": "string",
-            "description": "clangd path for compile_commands.json"
+            "description": "Path to the Compilation Database directory (containing compile_commands.json). " +
+                "This file contains compiler commands for each source files which are important for Clangd to properly analyze files. " +
+                "It can be generated by several tools such as CMake."
         },
-        "cpp.clangdPath": {
+        "cpp.clangdCommand": {
```

Comment:
The executable and the args should be provided through an env. It will allow to configure how to spawn a language server in a multi-container deployment. I.e. symmetrically to https://github.com/theia-ide/theia/blob/master/packages/terminal/src/node/shell-process.ts#L63

Reply:
I foresee that we will need to spawn different clangd (on different paths) depending on the configuration selected (debug, release, different architecture, etc). They can have different built-ins, different language extensions, etc. We don't have the concept of "configuration" in Theia yet but it will come. I'm not sure how to make that work nicely in the context of multi-container deployment. We need a way for the backend to spawn the correct clangd with the correct compilation database (and other settings) based on the active (last selected) configuration. To illustrate a bit the concepts, this would be a file (or preferences) tied to the workspace:
```
{
	"activeConfiguration": "Debug-X86",
	"configurations": [{
			"name": "Debug-X86",
			"serverCommand": "/path/to/normal/clangd",
			"serverCommandArgs": "-trace",
			"clangd.config": {
				"compilationDatabaseDirectory": "/path/containing/compilecommands",
				"index": {
					"exclusions": ["tests-special/*","powerpc/*"]
				}
			}
		},
		{
			"name": "Debug-PowerPC-with-special-language-extension",
			"serverCommand": "/path/to/special/clangd",
			"serverCommandArgs": "-trace",
			"clangd.config": {
				"compilationDatabaseDirectory": "/path/containing/specialcompilecommands",
				"index": {
					"exclusions": ["tests/*", "x86/*"]
				}
			}
		},
		{
			"name": "Debug-GCC?",
			"serverCommand": "/path/to/special/gccd?",
			"serverCommandArgs": "-v",
			"gcc.config": {
				"??": "??"
			}
		}
	]
}
```

We could move the clangd configs to a clangd-read file but there would be multiple per workspace. So it would look like:

```
{
	"activeConfiguration": "Debug-X86",
	"configurations": [{
			"name": "Debug-X86",
			"serverCommand": "/path/to/normal/clangd",
			"clangd.configFile": "/path/to/clangd.debug.x86.json"
		},
		{
			"name": "Debug-PowerPC-with-special-language-extension",
			"serverCommand": "/path/to/special/clangd",
			"clangd.configFile": "/path/to/clangd.debug.powerpc.ext.json"
		}
	]
}
```

We still need the backend to spawn the correct clangd and then clangd can read the clangd.json file. Either the front-end contains that configuration information and sends it to the backend (but then it cannot be shared between front-ends) or the backend has to read this information from the workspace, which is pretty much back to having backend preferences. With environment variables, it would be quite difficult, and in the above example there are only 2 configs:

THEIA_CPP_CONFIGS="Debug,Debug-PowerPC-with-special-language-extension"
THEIA_CPP_ACTIVE_CONFIG="Debug"
THEIA_CPP_CONFIG_DEBUG_X86_SERVER_COMMAND="/path/to/normal/clangd"
THEIA_CPP_CONFIG_DEBUG_X86_CLANGD_CONFIG_FILE="/path/to/clangd.debug.x86.json"
THEIA_CPP_CONFIG_POWERPC_SERVER_COMMAND="/path/to/special/clangd"
THEIA_CPP_CONFIG_POWERPC_CLANGD_CONFIG_FILE="/path/to/clangd.debug.powerpc.ext.json"
...there could be dozens more.
The user would also have to remember to change THEIA_CPP_ACTIVE_CONFIG whenever the config is changed at run-time otherwise next time Theia is started it will be wrong.

I think to unblock @epatpol it's OK if for now we use some environment variables but I feel like we'll have to revisit this in the future.